src/pymatgen/io/vasp/inputs.py

```diff
@@ -709,10 +710,12 @@ class BadPoscarWarning(UserWarning):
     """Warning class for bad POSCAR entries."""
 
 
-class Incar(dict, MSONable):
+class Incar(UserDict, MSONable):
```

Comment:
Subclassing `UserDict` would make our life much easier, otherwise some methods like `update/setdefault` would not call the overridden setter method, and I believe that's [the recommended way to create a custom dict](https://docs.python.org/3/library/collections.html#collections.UserDict):
> [UserDict](https://docs.python.org/3/library/collections.html#collections.UserDict): wrapper around dictionary objects for easier dict subclassing

Similar to: https://github.com/materialsproject/pymatgen/blob/ed8542d8ebf18c0fa49e6d87386a52072bc31559/src/pymatgen/io/lobster/inputs.py#L56

Perhaps we should consider write a custom and reusable key case-insensitive dict at some point instead of writing this over and over again.